lib/hanami/entity.rb

```diff
@@ -61,238 +51,83 @@ module Hanami
-      # @since 0.5.1
-      # @api private
-      def allowed_attribute_name?(name)
-        !instance_methods.include?(name)
-      end
-
-      protected
-
-      # @see Class#inherited
-      def inherited(subclass)
-        subclass.attributes(*attributes)
-        super
-      end
-    end
-
-    # Defines a generic, inefficient initializer, in case that the attributes
-    # weren't explicitly defined with `.attributes=`.
+    # Instantiate a new entity
     #
-    # @param attributes [Hash] a set of attribute names and values
+    # @param attributes [Hash,#to_h,NilClass] data to initialize the entity
     #
-    # @raise NoMethodError in case the given attributes are trying to set unknown
-    #   or private methods.
+    # @return [Hanami::Entity] the new entity instance
     #
     # @since 0.1.0
-    #
-    # @see .attributes
-    def initialize(attributes = {})
-      attributes.each do |k, v|
-        setter = "#{ k }="
-        public_send(setter, v) if respond_to?(setter)
-      end
+    def initialize(attributes = nil)
+      @attributes = Utils::Hash.new((attributes || {}).dup).symbolize!
+      freeze
     end
 
-    # Overrides the equality Ruby operator
+    # Entity ID
     #
-    # Two entities are considered equal if they are instances of the same class
-    # and if they have the same #id.
+    # @return [Object,NilClass] the ID, if present
     #
-    # @since 0.1.0
-    def ==(other)
-      self.class == other.class &&
-         self.id == other.id
+    # @since x.x.x
+    def id
+      attributes.fetch(:id, nil)
     end
 
-    # Return the hash of attributes
-    #
-    # @since 0.2.0
+    # Handle dynamic accessors
     #
-    # @example
-    #   require 'hanami/model'
-    #   class User
-    #     include Hanami::Entity
-    #     attributes :name
-    #   end
+    # If internal attributes set has the requested key, it returns the linked
+    # value, otherwise it raises a <tt>NoMethodError</tt>
     #
-    #   user = User.new(id: 23, name: 'Luca')
-    #   user.to_h # => { :id => 23, :name => "Luca" }
-    def to_h
-      Hash[attribute_names.map { |a| [a, read_attribute(a)] }]
+    # @since x.x.x
+    def method_missing(m)
+      attributes.fetch(m) { super }
     end
 
-    # Return the set of attribute names
+    # Implement generic equality for entities
+    #
+    # Two entities are equal if they are instances of the same class and they
+    # have the same id.
     #
-    # @since 0.5.1
+    # @param other [Object] the object of comparison
     #
-    # @example
-    #   require 'hanami/model'
-    #   class User
-    #     include Hanami::Entity
-    #     attributes :name
-    #   end
+    # @return [FalseClass,TrueClass] the result of the check
     #
-    #   user = User.new(id: 23, name: 'Luca')
-    #   user.attribute_names # #<Set: {:id, :name}>
-    def attribute_names
-      self.class.attributes
+    # @since 0.1.0
+    def ==(other)
+      self.class == other.class &&
+        id == other.id
```

Comment:
I'm not really familiar with DDD (any interesting read?), but what's the value in having a stale entity A be equal to itself after modification? (`A = UserRepository.create(...); B = UserRepository.modifyEmail(A.id, ...); A == B # => true`)